### ISSUE TITLE
Jackass The Games runs super slow

### ISSUE THREAD

**Issue by Panderner** [hrydgard/ppsspp#13676]
Jackass The Game runs super slow similar to Test Drive Unlimited.
![Screenshot_2020-11-18-15-42-46-60_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37503397/99611059-ece1c700-2a4d-11eb-969a-d59a8bd37cac.png)
It runs at 4-5 fps (5-6%) on Realme C2

**Comment by Panderner** [comment]
However it rans full speed on pc running windows on all backends:
![Screenshot (1)](https://user-images.githubusercontent.com/37503397/99668066-35c66980-2aa8-11eb-9c60-a4a2d8523f74.png)

Specs:
AMD A8-7650K Radeon R7
GTX 770
8GB RAM
Windows 10 Pro

**Comment by hrydgard** [comment]
[GE frame dump](https://github.com/hrydgard/ppsspp/wiki/How-to-create-a-frame-dump)?

**Comment by Panderner** [comment]
Yes Here's a GE Dump:
[recording.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/5566898/recording.ppdmp.zip)

**Comment by hrydgard** [comment]
Yeah this does the same thing as the Sonic games for color correction, but slightly differently - it actually textures directly from the framebuffer being rendered to, which (currently) forces PPSSPP to do "traditional" "depal" through a framebuffer copy. Not too bad on desktop, but gets horribly expensive on mobile tiler hardware :(

There are a few ways, more and less hacky, that we could fix this...  For now your best bet is probably playing at 1x rendering resolution to reduce the speed hit but it'll probably still be too slow.

**Comment by Panderner** [comment]
Even 1x resolution can't run this game on my phone.

**Comment by hrydgard** [comment]
well, you'll have to run it on PC until I get around to figuring out a way to fix it. Good to have a report and a nice dump to repro it though, very helpful!

**Comment by Panderner** [comment]
Also POCO M3 rans full speed on all backends but Realme C2 does not:
![Screenshot_2020-12-16-11-23-06-048_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/102304743-221ffd00-3f99-11eb-88f9-d9cab9ad3682.jpg)

**Comment by Panderner** [comment]
Performance comparison for this game:

POCO M3:
![Screenshot_2021-03-17-20-19-44-272_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/111466604-771d0c00-875e-11eb-8fea-c8acc0761f38.jpg)

Realme C2:
![Screenshot_2021-03-17-20-19-49-62_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37503397/111466748-a16ec980-875e-11eb-956d-77c0648ae8df.png)

**Comment by 71knight** [comment]
Come to find out the P22 processor is only half as powerful as a Snapdragon 662.

**Comment by Gamemulatorer** [comment]
This is a device issue not a ppsspp realme c2 is a weak phone unlike Poco M3 🤷

**Comment by Panderner** [comment]
> This is a device issue not a ppsspp realme c2 is a weak phone unlike Poco M3 🤷

@Gamemulatorer can you test this game on your phone? We need testing more devices if the game runs super slow.

**Comment by Gamemulatorer** [comment]
It's running smooth on my Redmi Note 9 both VK and OGL

**Comment by Panderner** [comment]
> It's running smooth on my Redmi Note 9 both VK and OGL

Oh ok

**Comment by unknownbrackets** [comment]
I think we could maybe more generically detect a situation where we have a:
 * Linear CLUT ramp
 * CLUT16/CLUT32 source with masking for a single color channel
 * RGBA write mask only writing the same channel
 * Through mode so we have coordinates (non-overlapping)
 * No other tests (depth, stencil, etc.) active

And queue the draws separately to combine as a single batch.  If they all had the same source texture and the same destination, it almost seems like we could ignore overlap checks in every practical case.

-[Unknown]

**Comment by hrydgard** [comment]
Not super easy to do that detection though, preventing flushing between the draws. 

To do that sort of thing easily, ideally we'd generate a higher level representation of the frame that we could postprocess (higher level than current renderqueues)....

**Comment by Panderner** [comment]
I tested Samsung Galaxy A12 (Not Mine) using vulkan.
![Screenshot_20210620-164628_PPSSPP](https://user-images.githubusercontent.com/37503397/122667842-99a04480-d1e7-11eb-8d96-87d9a23ff9c3.jpg)
It's a same gpu as my Realme C2

**Comment by Gamemulatorer** [comment]
This game is also super lag 🤣

**Comment by Panderner** [comment]
> This game is also super lag

@Gamemulatorer can you send me the screenshot?

**Comment by Gamemulatorer** [comment]
Maybe this game is only lag and graphics issue on Mali GPU?
![Screenshot_2021-08-23-17-33-31-645_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/130425347-7949f7d2-fbf3-48be-bd50-443706ec6272.jpg)
![Screenshot_2021-08-23-17-32-59-020_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/130425359-8978e552-4cdc-4484-b259-8cac36531952.jpg)
[jackass.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7030817/jackass.ppdmp.zip)

**Comment by Panderner** [comment]
Thank you @Gamemulatorer for testing this game for your Redmi Note 9. I also tested my Realme C2 which is using a PowerVR GPU is also running slow, My POCO M3 and my PC runs smoothly.

**Comment by hrydgard** [comment]
Yeah, it's well known that this game is running slowly. It's gonna need some special tricks like the Sonic games.

**Comment by Gamemulatorer** [comment]
> Yeah, it's well known that this game is running slowly. It's gonna need some special tricks like the Sonic games.

So this issues is related to this https://github.com/hrydgard/ppsspp/issues/10908

**Comment by hrydgard** [comment]
Yes

**Comment by hrydgard** [comment]
Gonna see if the Killzone workaround #15934 can be adapted for this.

**Comment by hrydgard** [comment]
Nope, this one is more like Sonic Rivals, although it's not using a temporary buffer, it's doing "depal" directly from the framebuffer to the same pixels on itself in little tiles, somehow that doesn't cause conflicts on the hardware. We end up depal-ing the entire texture, for every tiles, and there's a LOT of tiles.

Fortunately it's through-mode, we should be able to limit the region of the texture we depal for each draw fairly easily, and then maybe do a little post-rearrangement... 

So I think this can be improved substantially with some generic checks, but to really fly this will need some game-specific intervention similar to Sonic Rivals.

**Comment by Panderner** [comment]
My Realme C2 is broken unable to test this again hopely my Poco M3 runs ok for my phone this must be PowerVR GPU specific.

**Comment by Gamemulatorer** [comment]
> My Realme C2 is broken unable to test this again hopely my Poco M3 runs ok for my phone this must be PowerVR GPU specific.

This should improve now on your powervr gpu phone since there's a lot of improvement recently in powervr gpu https://github.com/hrydgard/ppsspp/pulls?q=is%3Apr+label%3A%22PowerVR+GPU%22+is%3Aclosed

**Comment by hrydgard** [comment]
Possibly, but the powervr improvements are mostly about speed of shader compilation and stuff like that. This game still runs very inefficiently and needs fixing.

**Comment by Gamemulatorer** [comment]
Except for the performance issue of this there's also a graphics issue below the screen.
HW Vulkan/OpenGL
![IMG_20230314_220736](https://user-images.githubusercontent.com/37603562/225027733-a4e698eb-8522-4013-a251-fdffa9546392.jpg)
SOFTWARE
![Screenshot_20230314_220544_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/225027792-0418930e-6318-40b6-8b63-960d5a6379a0.jpg)

**Comment by hrydgard** [comment]
New report from Discord, by "oh well": 

> I've tried to run it numerous times in the past, it never really works.
> Like it can load until the minigame menu, but when I try to play any of the first minigames (e.g. Pachinko Precipice), it gets stuck on the loading screen

**Comment by ISASLIMBOY28** [comment]
Run really slow on Tecno Spark 10 Pro (Helio G88/Mali-G52 MC2/8GB Ram). On Vulkan, got 10ish fps. On OpenGL, Got 20ish fps.

**Comment by hrydgard** [comment]
I'm aware that this is still not fixed. It's a tricky one, haven't gotten around to it - the potential fix idea I have is quite fiddly.

**Comment by Panderner** [comment]
It's seems this game is likely runs very slow on non adreno device.

**Comment by Panderner** [comment]
This game also runs slowly on my Samsung Galaxy A35 (Exynos 1380, Mali G68 MP5)
![Screenshot_20240914_095431_PPSSPP](https://github.com/user-attachments/assets/6d29ffea-55b8-4570-a94f-7d7c12398760)

**Comment by TAbdiukov** [comment]
I believe this issue is fixed in: https://github.com/hrydgard/ppsspp/issues/12083

If not, then [this comment of mine](https://github.com/hrydgard/ppsspp/issues/12083#issuecomment-1811350825) can be relevant (as a workaround),
> Don't use Direct3D 11 or OpenGL. Use Vulkan or Direct3D 9

(though Direct3D 9 has been deprecated now)
